kratos/containers/pointer_vector_set.h

```diff
@@ -558,54 +558,118 @@ class PointerVectorSet final
+    iterator insert(const TPointerType& value)
     {
-        ptr_iterator sorted_part_end;
-
-        key_type key = KeyOf(*pData);
-
-        if (mData.size() - mSortedPartSize >= mMaxBufferSize) {
-            Sort();
-            sorted_part_end = mData.end();
-        } else
-            sorted_part_end = mData.begin() + mSortedPartSize;
-
-        ptr_iterator i(std::lower_bound(mData.begin(), sorted_part_end, key, CompareKey()));
-        if (i == sorted_part_end) {
-            mSortedPartSize++;
-            return mData.insert(sorted_part_end, pData);
+        auto itr_pos = std::lower_bound(mData.begin(), mData.end(), KeyOf(*value), CompareKey());
+        if (itr_pos == mData.end()) {
+            // the position to insert is at the end.
+            mData.push_back(value);
+            mSortedPartSize = mData.size();
+            return iterator(mData.end() - 1);
+        } else if (EqualKeyTo(KeyOf(*value))(*itr_pos)) {
+            // already found existing element with the same key, hence returning the existing element.
+            return iterator(itr_pos);
+        } else {
+            // insert the new value before the itr_pos.
+            mSortedPartSize = mData.size() + 1;
+            return mData.insert(itr_pos, value);
         }
+    }
 
-        if (!EqualKeyTo(key)(*i))
-            if ((i = std::find_if(sorted_part_end, mData.end(), EqualKeyTo(key))) == mData.end()) {
-                mData.push_back(pData);
+    /**
+     * @brief Inserts a pointer at the specified position.
+     * @details This function inserts a given pointer. If the given position_hint is valid, then
+     * it uses that to insert the value, otherwise the position_hint is discarded to maintain the dataset
+     * sorted. If there is an existing element with the same key as in the value, then an iterator for
+     * the existing element is returned.
+     * @param position_hint An iterator pointing to the position where the pointer may be inserted.
+     * @param value The pointer to be inserted.
+     * @return An iterator pointing to the inserted element.
+     */
+    iterator insert(const_iterator position_hint, const TPointerType& value)
+    {
+        if (empty()) {
+            // the dataset is empty. So use push back.
+            mData.push_back(value);
+            mSortedPartSize = mData.size();
+            return iterator(mData.end() - 1);
+        } else if (position_hint == cend()) {
+            // trying to insert at the end.
+            if (KeyOf(*(position_hint - 1)) < KeyOf(*value)) {
+                // key at the position hint is less than the value of key. Hence position hint
+                // is valid. So using the push back.
+                mData.push_back(value);
+                mSortedPartSize = mData.size();
                 return iterator(mData.end() - 1);
+            } else {
+                // given position is invalid. Hence, discarding the hint.
+                return insert(value);
```

Comment:
Maybe a minor improvement, but I would not calling the insert here and just add the last statement of ifs in insert to avoid reevaluating for push_back

Reply:
done :)

Reply:
I see....